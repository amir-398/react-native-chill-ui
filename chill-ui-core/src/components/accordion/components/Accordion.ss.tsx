import type { AccordionPropsSs } from '@types';

import { BoxSs } from '@components/box';
import { PropsWithChildren } from 'react';

import styles from '../styles/Accordion.ss.styles';
import { AccordionProvider } from './AccordionContext';
import { accordionDefaultProps } from '../utils/defaultProps';

/**
 * Accordion component provides a collapsible content area with support for single or multiple items.
 *
 * @example
 * ```tsx
 * <Accordion type="single" collapsible>
 *   <AccordionItem value="item-1">
 *     <AccordionTrigger>Is it accessible?</AccordionTrigger>
 *     <AccordionContent>Yes. It adheres to the WAI-ARIA design pattern.</AccordionContent>
 *   </AccordionItem>
 * </Accordion>
 * ```
 *
 * @param children - Child AccordionItem components
 * @param type - Whether single or multiple items can be open at once ('single' | 'multiple')
 * @param collapseIcon - Icon to show when item is expanded (default: 'angle-down-solid')
 * @param collapsible - Whether items can be collapsed when type is 'single' (default: false)
 * @param defaultValue - Default open items (string for single, array for multiple)
 * @param disabled - Disable the entire accordion (default: false)
 * @param expandIcon - Icon to show when item is collapsed (default: 'angle-down-solid')
 * @param hasCollapseIcon - Whether to show collapse/expand icons (default: true)
 * @param iconPosition - Position of the expand/collapse icon (default: 'right')
 * @param onValueChange - Callback fired when accordion state changes
 * @param props - Additional ViewProps for the container
 *
 * @see {@link https://github.com/your-repo/chill-ui/tree/main/src/components/accordion/README.md Documentation}
 */
export default function Accordion(props: PropsWithChildren<AccordionPropsSs>) {
  const {
    children,
    collapseIcon = accordionDefaultProps.collapseIcon,
    collapsible = accordionDefaultProps.collapsible,
    defaultValue,
    disabled = accordionDefaultProps.disabled,
    expandIcon = accordionDefaultProps.expandIcon,
    hasCollapseIcon = accordionDefaultProps.hasCollapseIcon,
    iconPosition = accordionDefaultProps.iconPosition,
    onValueChange,
    style,
    type = accordionDefaultProps.type,
  } = props;

  return (
    <AccordionProvider
      type={type!}
      collapsible={collapsible}
      defaultValue={defaultValue}
      disabled={disabled}
      hasAnimation={false}
      animationDuration={0}
      hasCollapseIcon={hasCollapseIcon ?? true}
      iconPosition={iconPosition}
      expandIcon={expandIcon}
      collapseIcon={collapseIcon}
      onValueChange={onValueChange}
    >
      <BoxSs style={[styles.accordion, style]}>{children}</BoxSs>
    </AccordionProvider>
  );
}
