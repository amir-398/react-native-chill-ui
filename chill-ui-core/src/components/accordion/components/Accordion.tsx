import type { AccordionPropsTw } from '@types';

import { Box } from '@components/box';
import { PropsWithChildren } from 'react';
import { classNameHandler, cn, styleHandler } from '@utils';

import styles from '../styles/Accordion.styles';
import { AccordionProvider } from './AccordionContext';
import { accordionDefaultProps } from '../utils/defaultProps';
import { accordionClassName } from '../styles/Accordion.variants';
import { classNamePropsHandler } from '../../../utils/hybrid/classNamePropsHandler';

/**
 * Accordion component provides a collapsible content area with support for single or multiple items.
 * Automatically detects NativeWind availability and falls back to StyleSheet if needed.
 *
 * @example
 * ```tsx
 * <Accordion type="single" collapsible>
 *   <AccordionItem value="item-1">
 *     <AccordionTrigger>Is it accessible?</AccordionTrigger>
 *     <AccordionContent>Yes. It adheres to the WAI-ARIA design pattern.</AccordionContent>
 *   </AccordionItem>
 * </Accordion>
 * ```
 *
 * @param children - Child AccordionItem components
 * @param type - Whether single or multiple items can be open at once ('single' | 'multiple')
 * @param className - Custom CSS classes for styling
 * @param collapseIcon - Icon to show when item is expanded (default: 'angle-down-solid')
 * @param collapsible - Whether items can be collapsed when type is 'single' (default: false)
 * @param defaultValue - Default open items (string for single, array for multiple)
 * @param disabled - Disable the entire accordion (default: false)
 * @param expandIcon - Icon to show when item is collapsed (default: 'angle-down-solid')
 * @param hasCollapseIcon - Whether to show collapse/expand icons (default: true)
 * @param iconPosition - Position of the expand/collapse icon (default: 'right')
 * @param onValueChange - Callback fired when accordion state changes
 * @param props - Additional ViewProps for the container
 *
 * @see {@link https://github.com/your-repo/chill-ui/tree/main/src/components/accordion/README.md Documentation}
 */
export default function Accordion(props: PropsWithChildren<AccordionPropsTw>) {
  classNamePropsHandler(props, 'Accordion');
  const {
    children,
    className,
    collapseIcon = accordionDefaultProps.collapseIcon,
    collapsible = accordionDefaultProps.collapsible,
    defaultValue,
    disabled = accordionDefaultProps.disabled,
    expandIcon = accordionDefaultProps.expandIcon,
    hasCollapseIcon = accordionDefaultProps.hasCollapseIcon,
    iconPosition = accordionDefaultProps.iconPosition,
    onValueChange,
    style,
    type = accordionDefaultProps.type,
  } = props;

  return (
    <AccordionProvider
      type={type!}
      collapsible={collapsible}
      defaultValue={defaultValue}
      disabled={disabled}
      hasAnimation={false}
      animationDuration={0}
      hasCollapseIcon={hasCollapseIcon ?? true}
      iconPosition={iconPosition}
      expandIcon={expandIcon}
      collapseIcon={collapseIcon}
      onValueChange={onValueChange}
    >
      <Box
        {...classNameHandler(cn(accordionClassName, className))}
        {...styleHandler({ defaultStyle: styles.accordion, style })}
      >
        {children}
      </Box>
    </AccordionProvider>
  );
}
