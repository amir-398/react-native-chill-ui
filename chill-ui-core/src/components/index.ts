export * from './box';
export * from './toast';
export * from './loadingIndicatorsKit';
export * from './accordion';
export * from './dialog';
export * from './ripplePressable';
export * from './animatedBox';
export * from './icon';
export * from './avatar';
export * from './string';
export * from './inputDropdown';
export * from './AutocompleteDropdown';
export * from './highlightString';
export * from './buttonIcon';
export * from './checkbox';
export * from './chip';
export * from './maskedInput';
export * from './dropdownMenu';
export * from './inputSelectDropdown';
export * from './phoneNumberInput';
export * from './placesInput';
export * from './segmentedControl';
export * from './separator';
export * from './skeletons';
export * from './timePicker';
export * from './toggle';
export * from './wrapper';
