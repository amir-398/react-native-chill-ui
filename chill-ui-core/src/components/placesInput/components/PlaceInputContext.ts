import { AutocompleteDropdownContext } from '@components/AutocompleteDropdown/context/AutoCompleteDropdownContext';

export const PlaceInputContext = AutocompleteDropdownContext;
