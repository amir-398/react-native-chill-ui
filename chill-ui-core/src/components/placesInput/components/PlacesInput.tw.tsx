import { debounce } from '@utils';
import { Keyboard } from 'react-native';
import { useState, useEffect, useCallback, useMemo } from 'react';
import { AutocompleteDropdownTw } from '@components/AutocompleteDropdown';
import { PlacesInputPropsTw, PlaceTw, PlacesTw, PlacesResponseTw } from '@types';

/**
 * PlacesInput component that provides Google Places API integration for address autocomplete.
 * Features real-time search, debounced API calls, customizable rendering, and comprehensive place data.
 *
 * @example
 * ```tsx
 * // Basic usage with Google API key
 * <PlacesInput
 *   googleApiKey="your-google-api-key"
 *   placeholder="Enter your address"
 *   requiredCharactersBeforeSearch={3}
 *   requiredTimeBeforeSearch={500}
 *   clearQueryOnSelect={true}
 *   onSelect={(place) => {
 *     console.log('Selected place:', place.formattedAddress);
 *   }}
 *   onError={(error) => {
 *     console.error('Places API error:', error.message);
 *   }}
 *   onChangeText={(text) => console.log('Searching:', text)}
 * />
 * ```
 *
 * @param query - External query value to control the input
 * @param googleApiKey - Google Places API key (required)
 * @param queryCountries - Array of country codes to restrict search (e.g., ['US', 'CA'])
 * @param clearQueryOnSelect - Whether to clear the input after selection (default: false)
 * @param onSelect - Callback when a place is selected, receives Place or placePrediction object
 * @param onError - Callback when an error occurs during API calls, receives Error object
 * @param onOpenChange - Callback when open state changes
 * @param open - Controlled open state
 * @param defaultOpen - Default open state (uncontrolled)
 * @param requiredTimeBeforeSearch - Debounce delay in milliseconds before triggering search (default: 1000)
 * @param requiredCharactersBeforeSearch - Minimum characters before triggering search (default: 2)
 * @param selectedValue - Type of address component to display after selection:
 *   - 'longAddress': Full formatted address
 *   - 'shortAddress': Short formatted address
 *   - 'postal_code': Postal code only
 *   - 'locality': City/locality only
 *   - 'country': Country only
 *   - 'street_number': Street number only
 *   - 'route': Street name only
 * @param ...rest - All other props from AutocompleteDropdown component
 * @returns PlacesInput component with Google Places autocomplete
 */
export function PlacesInput(props: PlacesInputPropsTw) {
  const {
    clearQueryOnSelect = false,
    defaultOpen = false,
    googleApiKey,
    onChangeText,
    onError,
    onOpenChange,
    onSelect,
    open,
    query = '',
    queryCountries,
    requiredCharactersBeforeSearch = 2,
    requiredTimeBeforeSearch = 1000,
    selectedValue,
    ...rest
  } = props;
  const [localQuery, setLocalQuery] = useState(query);
  const [places, setPlaces] = useState<PlacesTw[]>([]);
  const [isLoading, setIsLoading] = useState(false);

  /**
   * Fetches places from Google Places API with debouncing
   * @param input - The search query text
   */
  const fetchPlaces = useCallback(
    async (input: string) => {
      if (!input || input.length < requiredCharactersBeforeSearch) {
        setPlaces([]);
        return;
      }
      setIsLoading(true);
      try {
        const requestData = {
          includedRegionCodes: queryCountries || [],
          input,
        };
        const response = await fetch('https://places.googleapis.com/v1/places:autocomplete', {
          body: JSON.stringify(requestData),
          headers: {
            'Content-Type': 'application/json',
            'X-Goog-Api-Key': googleApiKey,
            'X-Goog-FieldMask': [
              'suggestions.placePrediction.placeId',
              'suggestions.placePrediction.place',
              'suggestions.placePrediction.text.text',
            ].join(','),
          },
          method: 'POST',
        });

        if (!response.ok) {
          const errorData = await response.json().catch(() => ({}));
          const errorMessage = errorData.error?.message || `HTTP ${response.status}: ${response.statusText}`;
          if (onError) {
            onError(new Error(errorMessage));
          }
          setPlaces([]);
          return;
        }

        const data = (await response.json()) as PlacesResponseTw;
        if (data.error) {
          if (onError) {
            onError(data.error instanceof Error ? data.error : new Error(data.error.message || 'API Error'));
          }
          setPlaces([]);
          return;
        }

        setPlaces(data.suggestions || []);
      } catch (error) {
        if (onError) {
          onError(error instanceof Error ? error : new Error('Failed to fetch places'));
        }
        setPlaces([]);
      } finally {
        setIsLoading(false);
      }
    },
    [googleApiKey, queryCountries, requiredCharactersBeforeSearch, onError],
  );

  // Debounced function to fetch places with delay
  const debouncedFetchPlaces = useMemo(
    () => debounce(fetchPlaces, requiredTimeBeforeSearch),
    [fetchPlaces, requiredTimeBeforeSearch],
  );

  useEffect(() => {
    if (query !== localQuery) {
      setLocalQuery(query);
      debouncedFetchPlaces(query);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [query]);

  // Cleanup debounce on unmount
  useEffect(
    () => () => {
      debouncedFetchPlaces.cancel();
    },
    [debouncedFetchPlaces],
  );

  /**
   * Handles text input changes with debounced search
   * @param text - The input text value
   */
  const handleChangeText = useCallback(
    (text: string) => {
      setLocalQuery(text);
      debouncedFetchPlaces(text);
      if (onChangeText) onChangeText(text);
    },
    [debouncedFetchPlaces, onChangeText],
  );

  /**
   * Handles place selection and fetches detailed place information
   * @param item - The place item to select
   */
  const handlePlaceSelect = useCallback(
    async (item: PlacesTw) => {
      const id = item.placePrediction.placeId;
      const passedPlace = item.placePrediction;
      setIsLoading(true);
      try {
        const response = await fetch(`https://places.googleapis.com/v1/places/${id}`, {
          headers: {
            'Content-Type': 'application/json',
            'X-Goog-Api-Key': googleApiKey,
            'X-Goog-FieldMask': [
              'location',
              'formattedAddress',
              'postalAddress',
              'shortFormattedAddress',
              'addressComponents',
            ].join(','),
          },
          method: 'GET',
        });

        // Vérifier si la réponse HTTP est OK
        if (!response.ok) {
          const errorData = await response.json().catch(() => ({}));
          const errorMessage = errorData.error?.message || `HTTP ${response.status}: ${response.statusText}`;
          if (onError) {
            onError(new Error(errorMessage));
          }
          setLocalQuery(passedPlace.text.text);
          if (onSelect) onSelect(passedPlace);
          return;
        }

        const data = (await response.json()) as PlaceTw;

        setIsLoading(false);

        if (selectedValue) {
          if (selectedValue === 'longAddress') {
            setLocalQuery(data.formattedAddress);
          } else if (selectedValue === 'shortAddress') {
            setLocalQuery(data.shortFormattedAddress);
          } else if (selectedValue === 'postal_code') {
            setLocalQuery(
              data.addressComponents.find(component => component.types.includes('postal_code'))?.longText || '',
            );
          } else if (selectedValue === 'locality') {
            setLocalQuery(
              data.addressComponents.find(component => component.types.includes('locality'))?.longText || '',
            );
          } else if (selectedValue === 'country') {
            setLocalQuery(
              data.addressComponents.find(component => component.types.includes('country'))?.longText || '',
            );
          } else if (selectedValue === 'street_number') {
            setLocalQuery(
              data.addressComponents.find(component => component.types.includes('street_number'))?.longText || '',
            );
          } else if (selectedValue === 'route') {
            setLocalQuery(data.addressComponents.find(component => component.types.includes('route'))?.longText || '');
          } else {
            setLocalQuery(
              data.addressComponents.find(component => component.types.includes(selectedValue))?.longText || '',
            );
          }
        } else {
          setLocalQuery(clearQueryOnSelect ? '' : (data && data.formattedAddress) || data.shortFormattedAddress);
        }
        if (onSelect) onSelect(data);
        Keyboard.dismiss();
      } catch (error) {
        setIsLoading(false);
        setLocalQuery(passedPlace.text.text);
        if (onError) {
          onError(error instanceof Error ? error : new Error('Failed to fetch place details'));
        }
        if (onSelect) onSelect(passedPlace);
      }
    },
    [clearQueryOnSelect, googleApiKey, onError, onSelect, selectedValue],
  );

  return (
    <AutocompleteDropdownTw
      {...rest}
      dataSet={places}
      valueField="placePrediction.text.text"
      onChangeText={handleChangeText}
      hasPerformSearch={false}
      onSelectItem={handlePlaceSelect}
      isLoading={isLoading}
      open={open}
      onOpenChange={onOpenChange}
      defaultOpen={defaultOpen}
    />
  );
}

PlacesInput.displayName = 'PlacesInput';
