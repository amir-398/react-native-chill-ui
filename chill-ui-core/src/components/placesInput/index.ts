export { PlacesInput } from './components/PlacesInput';
export { PlacesInput as PlacesInputSs } from './components/PlacesInput.ss';
export { PlacesInput as PlacesInputTw } from './components/PlacesInput.tw';
export { PlaceInputContext } from './components/PlaceInputContext';
