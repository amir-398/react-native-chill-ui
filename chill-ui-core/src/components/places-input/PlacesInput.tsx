/* eslint-disable react/prop-types */
import { Keyboard } from 'react-native';
import { useState, useEffect, useRef, useCallback } from 'react';

import { AutocompleteDropdown } from '../AutocompleteDropdown';
import { Place, Places, PlacesInputProps, PlacesResponse } from '../../types/placesInput.types';

/**
 * PlacesInput component that provides Google Places API integration for address autocomplete.
 * Features real-time search, debounced API calls, customizable rendering, and comprehensive place data.
 *
 * @example
 * ```tsx
 * // Basic usage with Google API key
 * <PlacesInput
 *   googleApiKey="your-google-api-key"
 *   onSelect={(place) => console.log(place)}
 * />
 *
 * // With country restriction and custom selection value
 * <PlacesInput
 *   googleApiKey="your-google-api-key"
 *   queryCountries={['US', 'CA']}
 *   selectedValue="locality"
 *   onSelect={(place) => {
 *     console.log('Selected place:', place.formattedAddress);
 *   }}
 * />
 *
 * // With custom validation and styling
 * <PlacesInput
 *   googleApiKey="your-google-api-key"
 *   placeholder="Enter your address"
 *   requiredCharactersBeforeSearch={3}
 *   requiredTimeBeforeSearch={500}
 *   clearQueryOnSelect={true}
 *   onSelect={(place) => {
 *     console.log('Selected place:', place.formattedAddress);
 *   }}
 *   onChangeText={(text) => console.log('Searching:', text)}
 * />
 * ```
 *
 * @param query - External query value to control the input
 * @param googleApiKey - Google Places API key (required)
 * @param queryCountries - Array of country codes to restrict search (e.g., ['US', 'CA'])
 * @param clearQueryOnSelect - Whether to clear the input after selection (default: false)
 * @param onSelect - Callback when a place is selected, receives Place or placePrediction object
 * @param requiredTimeBeforeSearch - Debounce delay in milliseconds before triggering search (default: 1000)
 * @param requiredCharactersBeforeSearch - Minimum characters before triggering search (default: 2)
 * @param selectedValue - Type of address component to display after selection:
 *   - 'longAddress': Full formatted address
 *   - 'shortAddress': Short formatted address
 *   - 'postal_code': Postal code only
 *   - 'locality': City/locality only
 *   - 'country': Country only
 *   - 'street_number': Street number only
 *   - 'route': Street name only
 * @param ...rest - All other props from AutocompleteDropdown component
 * @returns PlacesInput component with Google Places autocomplete
 */
function PlacesInput(props: PlacesInputProps) {
  const {
    clearQueryOnSelect = false,
    googleApiKey,
    onChangeText,
    onSelect,
    query = '',
    queryCountries,
    requiredCharactersBeforeSearch = 2,
    requiredTimeBeforeSearch = 1000,
    selectedValue,
    ...rest
  } = props;
  const [localQuery, setLocalQuery] = useState(query);
  const [places, setPlaces] = useState<Places[]>([]);
  const [isLoading, setIsLoading] = useState(false);
  const timeoutRef = useRef<NodeJS.Timeout | null>(null);

  // Cleanup du timeout au unmount
  useEffect(
    () => () => {
      if (timeoutRef.current) clearTimeout(timeoutRef.current);
    },
    [],
  );

  /**
   * Fetches places from Google Places API with debouncing
   * @param input - The search query text
   */
  const fetchPlaces = useCallback(
    async (input: string) => {
      if (!input || input.length < requiredCharactersBeforeSearch) {
        setPlaces([]);
        return;
      }
      setIsLoading(true);
      try {
        const requestData = {
          includedRegionCodes: queryCountries || [],
          input,
        };
        const response = await fetch('https://places.googleapis.com/v1/places:autocomplete', {
          body: JSON.stringify(requestData),
          headers: {
            'Content-Type': 'application/json',
            'X-Goog-Api-Key': googleApiKey,
            'X-Goog-FieldMask': [
              'suggestions.placePrediction.placeId',
              'suggestions.placePrediction.place',
              'suggestions.placePrediction.text.text',
            ].join(','),
          },
          method: 'POST',
        });
        const data = (await response.json()) as PlacesResponse;
        setPlaces(data.suggestions || []);
      } finally {
        setIsLoading(false);
      }
    },
    [googleApiKey, queryCountries, requiredCharactersBeforeSearch],
  );

  /**
   * Debounced function to fetch places with delay
   * @param input - The search query text
   */
  const debouncedFetchPlaces = useCallback(
    (input: string) => {
      if (timeoutRef.current) clearTimeout(timeoutRef.current);
      timeoutRef.current = setTimeout(() => {
        fetchPlaces(input);
      }, requiredTimeBeforeSearch);
    },
    [fetchPlaces, requiredTimeBeforeSearch],
  );

  // RafraÃ®chir la query externe
  useEffect(() => {
    if (query !== localQuery) {
      setLocalQuery(query);
      debouncedFetchPlaces(query);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [query]);

  /**
   * Handles text input changes with debounced search
   * @param text - The input text value
   */
  const handleChangeText = useCallback(
    (text: string) => {
      setLocalQuery(text);
      debouncedFetchPlaces(text);
      if (onChangeText) onChangeText(text);
    },
    [debouncedFetchPlaces, onChangeText],
  );

  /**
   * Handles place selection and fetches detailed place information
   * @param item - The place item to select
   */
  const handlePlaceSelect = useCallback(
    async (item: Places) => {
      const id = item.placePrediction.placeId;
      const passedPlace = item.placePrediction;
      setIsLoading(true);
      try {
        const response = await fetch(`https://places.googleapis.com/v1/places/${id}`, {
          headers: {
            'Content-Type': 'application/json',
            'X-Goog-Api-Key': googleApiKey,
            'X-Goog-FieldMask': [
              'location',
              'formattedAddress',
              'postalAddress',
              'shortFormattedAddress',
              'addressComponents',
            ].join(','),
          },
          method: 'GET',
        });
        const data = (await response.json()) as Place;

        setIsLoading(false);

        if (selectedValue) {
          if (selectedValue === 'longAddress') {
            setLocalQuery(data.formattedAddress);
          } else if (selectedValue === 'shortAddress') {
            setLocalQuery(data.shortFormattedAddress);
          } else if (selectedValue === 'postal_code') {
            setLocalQuery(
              data.addressComponents.find(component => component.types.includes('postal_code'))?.longText || '',
            );
          } else if (selectedValue === 'locality') {
            setLocalQuery(
              data.addressComponents.find(component => component.types.includes('locality'))?.longText || '',
            );
          } else if (selectedValue === 'country') {
            setLocalQuery(
              data.addressComponents.find(component => component.types.includes('country'))?.longText || '',
            );
          } else if (selectedValue === 'street_number') {
            setLocalQuery(
              data.addressComponents.find(component => component.types.includes('street_number'))?.longText || '',
            );
          } else if (selectedValue === 'route') {
            setLocalQuery(data.addressComponents.find(component => component.types.includes('route'))?.longText || '');
          } else {
            setLocalQuery(
              data.addressComponents.find(component => component.types.includes(selectedValue))?.longText || '',
            );
          }
        } else {
          setLocalQuery(clearQueryOnSelect ? '' : (data && data.formattedAddress) || data.shortFormattedAddress);
        }
        if (onSelect) onSelect(data);
        Keyboard.dismiss();
      } catch {
        setIsLoading(false);
        setLocalQuery(passedPlace.text.text);
        if (onSelect) onSelect(passedPlace);
      }
    },
    [clearQueryOnSelect, googleApiKey, onSelect, selectedValue],
  );

  return (
    <AutocompleteDropdown
      {...rest}
      dataSet={places}
      valueField="placePrediction.text.text"
      onChangeText={handleChangeText}
      hasPerformSearch={false}
      onSelectItem={handlePlaceSelect}
      isLoading={isLoading}
    />
  );
}

export default PlacesInput;
