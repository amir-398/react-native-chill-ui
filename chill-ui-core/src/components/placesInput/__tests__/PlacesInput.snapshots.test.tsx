import { render } from '@testing-library/react-native';

import { PlacesInput } from '../components/PlacesInput';

// Mocks
jest.mock('../../../utils', () => ({
  classNameHandler: jest.fn(() => ({})),
  classNamePropsHandler: jest.fn(),
  cn: jest.fn((...args) => args.filter(Boolean).join(' ')),
  colorVariantPropsHandler: jest.fn(),
  debounce: jest.fn(fn => {
    const debouncedFn = jest.fn(fn);
    debouncedFn.cancel = jest.fn();
    return debouncedFn;
  }),
  isString: jest.fn(value => typeof value === 'string'),
  styleHandler: jest.fn(() => ({})),
}));

jest.mock('../../../components/AutocompleteDropdown', () => {
  const React = require('react');
  const { Text, TextInput, TouchableOpacity, View } = require('react-native');

  function MockAutocompleteDropdown({ children, onChangeText, onSelectItem, ...props }: any) {
    return (
      <View testID="autocomplete-dropdown" {...props}>
        <TextInput
          testID="places-input"
          onChangeText={onChangeText}
          placeholder={props.inputProps?.placeholder || 'Enter your address'}
        />
        {children}
        <TouchableOpacity
          testID="select-place"
          onPress={() =>
            onSelectItem &&
            onSelectItem({
              placePrediction: {
                placeId: 'test-place-id',
                text: { text: 'Test Address' },
              },
            })
          }
        >
          <Text>Select Place</Text>
        </TouchableOpacity>
      </View>
    );
  }

  return {
    AutocompleteDropdown: MockAutocompleteDropdown,
    AutocompleteDropdownSs: MockAutocompleteDropdown,
    AutocompleteDropdownTw: MockAutocompleteDropdown,
  };
});

// Mock fetch globally
global.fetch = jest.fn();

describe('PlacesInput Snapshots', () => {
  const mockGoogleApiKey = 'test-api-key';

  beforeEach(() => {
    jest.clearAllMocks();
    (global.fetch as jest.Mock).mockClear();
  });

  it('should match snapshot with basic props', () => {
    const tree = render(<PlacesInput googleApiKey={mockGoogleApiKey} />);
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot with placeholder', () => {
    const tree = render(
      <PlacesInput googleApiKey={mockGoogleApiKey} inputProps={{ placeholder: 'Search for a location' }} />,
    );
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot with country restrictions', () => {
    const tree = render(<PlacesInput googleApiKey={mockGoogleApiKey} queryCountries={['US', 'CA', 'GB']} />);
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot with custom styling', () => {
    const tree = render(
      <PlacesInput
        googleApiKey={mockGoogleApiKey}
        className="custom-class"
        style={{ backgroundColor: '#fff', margin: 16 }}
      />,
    );
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot with all props combined', () => {
    const tree = render(
      <PlacesInput
        googleApiKey={mockGoogleApiKey}
        queryCountries={['US', 'CA']}
        selectedValue="locality"
        clearQueryOnSelect
        requiredCharactersBeforeSearch={3}
        requiredTimeBeforeSearch={500}
        defaultOpen={false}
        inputProps={{
          placeholder: 'Enter your full address',
          size: 'md',
        }}
      />,
    );
    expect(tree).toMatchSnapshot();
  });
});
