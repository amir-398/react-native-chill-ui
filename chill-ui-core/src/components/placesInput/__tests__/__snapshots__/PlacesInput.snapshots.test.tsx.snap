// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlacesInput Snapshots should match snapshot with all props combined 1`] = `
<View
  dataSet={[]}
  defaultOpen={false}
  hasPerformSearch={false}
  inputProps={
    {
      "placeholder": "Enter your full address",
      "size": "md",
    }
  }
  isLoading={false}
  testID="autocomplete-dropdown"
  valueField="placePrediction.text.text"
>
  <TextInput
    onChangeText={[Function]}
    placeholder="Enter your full address"
    testID="places-input"
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="select-place"
  >
    <Text>
      Select Place
    </Text>
  </View>
</View>
`;

exports[`PlacesInput Snapshots should match snapshot with basic props 1`] = `
<View
  dataSet={[]}
  defaultOpen={false}
  hasPerformSearch={false}
  isLoading={false}
  testID="autocomplete-dropdown"
  valueField="placePrediction.text.text"
>
  <TextInput
    onChangeText={[Function]}
    placeholder="Enter your address"
    testID="places-input"
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="select-place"
  >
    <Text>
      Select Place
    </Text>
  </View>
</View>
`;

exports[`PlacesInput Snapshots should match snapshot with country restrictions 1`] = `
<View
  dataSet={[]}
  defaultOpen={false}
  hasPerformSearch={false}
  isLoading={false}
  testID="autocomplete-dropdown"
  valueField="placePrediction.text.text"
>
  <TextInput
    onChangeText={[Function]}
    placeholder="Enter your address"
    testID="places-input"
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="select-place"
  >
    <Text>
      Select Place
    </Text>
  </View>
</View>
`;

exports[`PlacesInput Snapshots should match snapshot with custom styling 1`] = `
<View
  className="custom-class"
  dataSet={[]}
  defaultOpen={false}
  hasPerformSearch={false}
  isLoading={false}
  style={
    {
      "backgroundColor": "#fff",
      "margin": 16,
    }
  }
  testID="autocomplete-dropdown"
  valueField="placePrediction.text.text"
>
  <TextInput
    onChangeText={[Function]}
    placeholder="Enter your address"
    testID="places-input"
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="select-place"
  >
    <Text>
      Select Place
    </Text>
  </View>
</View>
`;

exports[`PlacesInput Snapshots should match snapshot with placeholder 1`] = `
<View
  dataSet={[]}
  defaultOpen={false}
  hasPerformSearch={false}
  inputProps={
    {
      "placeholder": "Search for a location",
    }
  }
  isLoading={false}
  testID="autocomplete-dropdown"
  valueField="placePrediction.text.text"
>
  <TextInput
    onChangeText={[Function]}
    placeholder="Search for a location"
    testID="places-input"
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="select-place"
  >
    <Text>
      Select Place
    </Text>
  </View>
</View>
`;
